File: src/main/python/datastructures/linkedlist/linked_list.py
class Node:
    """Linked List Node that contains data and next node."""
    def __init__(self, data = None, next_ = None):
        self.data = data
        self.next = next_

    def __str__(self):
        return str(self.data)

class LinkedList:
    """Single linked list using nodes.
    
    __init__ method creates a sentinel node and initializes the length of the list to 0.
    
    Attributes:
        top_sentinel (Node): empty node that is always at the beginning of the linked list.
        length (int): size of the linked list. 
    """

    def __init__(self):
        self.top_sentinel = Node()
        self.length = 0

        
    def __iter__(self):
        node = self.top_sentinel.next
        while node is not None:
            yield node.data
            node = node.next
        
    def __str__(self):
        return str(list(self))
    
    def is_empty(self):
        """Return true if the list is empty, false otherwise."""
        return self.length == 0
        
    def size(self):
        """Return the size of the linked list."""
        return self.length
    
    def add_first(self, data):
        """Add a node to the top of the linked list."""
        if self.is_empty() == True:
            # if the list is empty then have the top sentinel point to the given node.
            self.top_sentinel.next = Node(data)
        else:
            # if the list is not empty, create a new node and have it point to the old top node.
            new_node = Node(data, self.top_sentinel.next)
            self.top_sentinel.next = new_node
    
        self.length += 1
        
    # add has same functionality as add_first.
    add = add_first
    
    
    

    
    
    

File: src/main/ruby/datastructures/linkedlist/linked_list.rb
# Sentinel Linked list representation using nodes. 
# insert/delete at beginning: O(1), insert/delete at end: O(n)

class Node
    attr_accessor :data, :next

    def initialize(data = nil, next_ = nil)
        @data = data
        @next = next_
    end

    def to_s
        "#{data}"
    end
end

class LinkedList
    def initialize
        @top_sentinel = Node.new
        @length = 0
    end

    def size
        @length
    end

    def is_empty
        @length == 0
    end 

    def push(data)
        if is_empty
            @top_sentinel.next = Node.new(data)
            
        else
            new_node = Node.new(data, @top_sentinel.next)
            @top_sentinel.next = new_node
        end
        @length += 1
    end

    def pop
        if is_empty
            raise IndexError.new "pop from empty list"
        elsif size == 1
            top_node = @top_sentinel.next
            @top_sentinel.next = nil
            return top_node.data
        else
            top_node = @top_sentinel.next
            @top_sentinel.next = top_node.next
            return top_node.data
        end
    end 
        
    def contains(target)
        return false if is_empty

        # return top node if the target 
        node = @top_sentinel.next
        return true if node.data == target
        
        #loop through list and return target if found
        while (node.next != nil)
            node = node.next
            return true if node.data == target
        end 
        
        #otherwise, not in list
        return false
    end
end
